usr/lib/soft_stdll/soft_specific.c

```diff
@@ -3051,3 +3075,540 @@ CK_RV token_specific_aes_cmac(STDLL_TokData_t *tokdata, CK_BYTE *message,
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (!ECDSA_SIG_set0(sig, r, s)) {
+        TRACE_ERROR("ECDSA_SIG_set0 failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    rc = ECDSA_do_verify(in_data, in_data_len, sig, ec_key);
+    switch (rc) {
+    case 0:
+        rc = CKR_SIGNATURE_INVALID;
+        break;
+    case 1:
+        rc = CKR_OK;
+        break;
+    default:
+        rc = CKR_FUNCTION_FAILED;
+        break;
+    }
+
+out:
+    if (sig != NULL)
+        ECDSA_SIG_free(sig);
+    if (ec_key != NULL)
+        EC_KEY_free(ec_key);
+
+    return rc;
+}
+
+CK_RV token_specific_ecdh_pkcs_derive(STDLL_TokData_t *tokdata,
+                                      CK_BYTE *priv_bytes,
+                                      CK_ULONG priv_length,
+                                      CK_BYTE *pub_bytes,
+                                      CK_ULONG pub_length,
+                                      CK_BYTE *secret_value,
+                                      CK_ULONG *secret_value_len,
+                                      CK_BYTE *oid, CK_ULONG oid_length)
+{
+    EC_KEY *ec_pub = NULL, *ec_priv = NULL;
+    CK_ULONG privlen;
+    int secret_len;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+
+    rc = make_ec_key_from_params(oid, oid_length, &ec_priv);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("make_ec_key_from_params failed\n");
+        goto out;
+    }
+
+    rc = fill_ec_key_from_privkey(ec_priv, priv_bytes, priv_length);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("fill_ec_key_from_privkey failed\n");
+        goto out;
+    }
+
+    rc = make_ec_key_from_params(oid, oid_length, &ec_pub);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("make_ec_key_from_params failed\n");
+        goto out;
+    }
+
+    rc = fill_ec_key_from_pubkey(ec_pub, pub_bytes, pub_length);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("fill_ec_key_from_privkey failed\n");
+        goto out;
+    }
+
+    privlen = (EC_GROUP_order_bits(EC_KEY_get0_group(ec_priv)) + 7) / 8;
+
+    secret_len = ECDH_compute_key(secret_value, privlen,
+                                  EC_KEY_get0_public_key(ec_pub), ec_priv,
+                                  NULL);
+    if (secret_len <= 0) {
+        TRACE_DEVEL("ECDH_compute_key failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        *secret_value_len = 0;
+        goto out;
+    }
+
+    *secret_value_len = secret_len;
+
+out:
+    if (ec_priv != NULL)
+        EC_KEY_free(ec_priv);
+    if (ec_pub != NULL)
+        EC_KEY_free(ec_pub);
+
+    return rc;
+}
+
+#endif
+
+CK_RV token_specific_object_add(STDLL_TokData_t * tokdata, SESSION * sess,
+                                OBJECT * obj)
+{
+    CK_ATTRIBUTE *attr = NULL;
+    CK_KEY_TYPE keytype;
+#ifndef NO_EC
+    EC_KEY *ec_key = NULL;
+    CK_RV rc;
+#endif
+
+    UNUSED(tokdata);
+    UNUSED(sess);
+
+    if (template_attribute_find(obj->template, CKA_KEY_TYPE, &attr) == FALSE)
+        return CKR_OK;
```

Comment:
Trace entry?

Reply:
No, this exit is not an error. It is just indicates that the object that is being added is not a key (and thus must not be handled here). token_specific_object_add is called by common code for ALL objects, regardless of its type.